### ISSUE TITLE
Cannot send JSON payload with logging label when executed in Cloud Function environment

### ISSUE THREAD

**Issue by ianyoung** [googleapis/python-logging#626]
I'm trying to log a message from a Cloud Function using the Python logging library which contains a string, a JSON payload, and a logging label. E.g:

```python
image_metrics = {
    "total_images_sourced": total_images_sourced,
    "total_images": total_images
}

log.info("image_metrics", extra={
    "labels": {"type": "image"},
    "json_fields": image_metrics  
})
```

I'm following the documented guidelines in [logging a JSON payload](https://googleapis.dev/python/logging/latest/std-lib-integration.html#logging-json-payloads) via `extra` using the `json_fields` key. And also [passing in a custom label](https://googleapis.dev/python/logging/latest/std-lib-integration.html#manual-metadata-using-the-extra-argument) also via `extra`.

I've integrated Cloud Logging successfully and when I run my Python script locally via the Functions Framework or directly via script execution, I can successfully see the logging statements appear as expected in Cloud Logging:

![gcl-local-success (1)](https://user-images.githubusercontent.com/66717/189708871-7b3d717f-178f-491f-8b6a-cee7a8597661.png)

However, when I deploy the Cloud Function with exactly the same code I see the following in Cloud Logging:

![gcl-cloud-function-failure](https://user-images.githubusercontent.com/66717/189704480-205f657b-dab8-4464-bcb9-42a2a1b49b6c.png)

The same log entry seems to be interpreted as a text payload instead of JSON payload, my label is ignored, and the name of my log doesn't get set.

I have tried the following test cases to see if I can get anything to work:

```python
# Test 1: JSON payload only
log.info(json.dumps(image_metrics))

# Test 2: Message + JSON payload via `extra` argument.
log.info("test_2", extra={"json_fields": image_metrics})

# Test 3: Message + Label via `extra` argument.
log.info("test_3", extra={"labels": {"type": "image"}})
```

Test 1 is the only test to work successfully, outputting the following:

![gcl-jsonpayload-only-success](https://user-images.githubusercontent.com/66717/189707676-4f2a6e57-8134-44fc-b34d-e72316e978e4.png)

So it seems that when run as a Cloud Function, I'm only able to log a message OR a JSON payload but not both and never a label. My log name also never appears to get picked up.

I may be wrong but I was expecting the same result when run in a Cloud Function environment that I get when running the application locally.

#### Environment details

  - Python version: 3.10.6
  - `google-cloud-logging` version: 3.2.2
  - Cloud Function: 1st gen

**Comment by ianyoung** [comment]
I also came across this issue which sounds like the same problem:
https://github.com/GoogleCloudPlatform/python-docs-samples/issues/5036

And this SO post:
https://stackoverflow.com/questions/64935095/writing-structured-json-log-messages-from-python-google-cloud-function

Neither of which appear to have been resolved.

**Comment by daniel-sanche** [comment]
I just tested your code, and it seems to work for me (using Python 3.8, GCF 1st gen)

![image](https://user-images.githubusercontent.com/3914119/189772311-73bc83dc-dfc6-42e4-9b68-a3b7e4d44b38.png)

My guess is that you didn't run the setup code in your functions environment before writing logs:
```
    client = google.cloud.logging.Client()
    client.setup_logging()
```

Without running the setup line, your logs will be output using GCF's built-in LogHandlers, and bypass the logging library's StructuredLogHandler, which is what manages the structured JSON formatting. I tried running the code without the setup line in place, and I got results in line with what you reported

**Comment by daniel-sanche** [comment]
Feel free to re-open if you think that's not the issue, and I can try to look again

**Comment by ianyoung** [comment]
Thanks for looking into it @daniel-sanche. You're right, I haven't included `client.setup_logging()`.

I currently have a logging module which contains:

```python
import logging

import google.cloud.logging
from google.cloud.logging.handlers import CloudLoggingHandler

def get_stream_handler():
    """Add a stream handler to log messages to the console."""
    stream_handler = logging.StreamHandler()
    stream_handler.setLevel(logging.INFO)
    return stream_handler


def get_gcloud_handler(log_name):
    """Add logging handler for GCP Cloud Logging."""
    gcloud_logging_client = google.cloud.logging.Client()
    gcloud_logging_handler = CloudLoggingHandler(
        gcloud_logging_client, name=log_name
    )
    return gcloud_logging_handler
    

def get_logger(log_name):
    logger = logging.getLogger(log_name)
    logger.setLevel(logging.DEBUG)

    # Make sure not to create multiple handlers 
    # or logging entries will appear multiple times.
    if not logger.hasHandlers():
        logger.addHandler(get_gcloud_handler(log_name))
        logger.addHandler(get_stream_handler())

    return logger
```

And in modules I use where I need to add logging I include:

```python
from app.services.logging import get_logger
log = get_logger('my_log_name')
```

I've setup two handlers so that I can log output to both the console and to Cloud Logging. In this configuration, how is it best to include `client.setup_logging()`?

**Comment by daniel-sanche** [comment]
One of the benefits of using `setup_logging` is that it can [choose the right Handler for you](https://googleapis.dev/python/logging/3.2.2/std-lib-integration.html#automatic-configuration) based on where your code is running, you typically shouldn't have to manage handlers yourself. 

The Handlers you are using in your code are:
- [`CloudLoggingHandler`](https://googleapis.dev/python/logging/3.2.2/handlers-cloud-logging.html), which sends logs in batches over the network. This Handler doesn't always work well with Cloud Functions, because the serverless environment may spin down quickly before the handler has a chance to send out the logs
- [`StreamHandler`](https://docs.python.org/3/library/logging.handlers.html#streamhandler), which isn't one of ours, it is a built-in Python handler. It will output your messages to standard out, but not in the [structured JSON format](https://cloud.google.com/functions/docs/monitoring/logging) that is parseable by GCP Cloud Logging. The logs output from this handler will show up like you reported: missing JSON fields, labels, severity, etc

Using these two handlers, you will likely see duplicated logs in Log Explorer, with the `StreamHandler` logs being unformatted, and `CloudLoggingHandler` logs being unreliable

The Handler we recommend for Cloud Functions is:
- [`StructuredLoggingHandler`](https://googleapis.dev/python/logging/3.2.2/handlers-structured-log.html), which prints properly formatted structured JSON log entries into standard out, which are then parsed and transferred into Cloud Logging by the GCP logging agents

Basically, we recommend just running `client.setup_logging()`, and letting it automatically choose the best handler for the environment the code is running in. But if you still want to set a handler manually, I hope this helps clarify the different use cases of each one

**Comment by ianyoung** [comment]
Thanks for the explanation. When adding `setup_logging(gcloud_logging_handler)` i do see duplicate log messages, but not without.

The only reason I went the route of adding custom handlers was because with `client.setup_logging()` i wasn't seeing any log messages in the console when developing locally. Is there a way to solve this?

**Comment by ianyoung** [comment]
It works! It seems I went about things the hard way. :)

It all makes much more sense now. I was thinking along the lines of needing to treat Cloud Logging as a regular handler to attach to the root logger. The differing results between local and the Cloud Functions environment had me confused.

Following your suggestion I've gone with the following:

```python
import logging

client = google.cloud.logging.Client()
client.setup_logging()

log = logging.getLogger()

if __name__ == '__main__':
    # Add a stream handler to log messages to the console.
    stream_handler = logging.StreamHandler()
    stream_handler.setLevel(logging.INFO)
    log.addHandler(stream_handler)
```

Since Cloud Functions don't ever trigger that block it makes for a handy place to store things that should only run when triggered locally in development.

I've given things a test and I finally have log messages output from Cloud Functions the same way that they do when run locally. And I still have console logging!

Thanks for your help on testing this and explaining the problem. Very much appreciated

**Comment by daniel-sanche** [comment]
Awesome, glad you got it working! Let me know if you run into any other issues using the library
